# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/test/xpcshell/common/head.js
# Commit: 46a5babaafd2
# Full Hash: 46a5babaafd2c0f2198f353d62254cf518dab391
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-15 04:16:11
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Implement initialization of entire persistent storage; r=dom-storage-reviewers,jari
#   
#   IndexedDB still calls QuotaManager::EnsurePersistentOriginIsInitializedInternal
#   from its idle database maintenance. However, we want to get rid of that so the
#   method can be made private. The idle database maintenance should only rely on
# ==============================================================================

diff -r 2ff5f3aaf841 -r 46a5babaafd2 dom/quota/test/xpcshell/common/head.js
--- a/dom/quota/test/xpcshell/common/head.js	Mon Oct 14 10:00:26 2024 +0000
+++ b/dom/quota/test/xpcshell/common/head.js	Mon Oct 14 10:06:14 2024 +0000
@@ -131,6 +131,13 @@
   return request;
 }
 
+function persistentStorageInitialized(callback) {
+  let request = SpecialPowers._getQuotaManager().persistentStorageInitialized();
+  request.callback = callback;
+
+  return request;
+}
+
 function temporaryStorageInitialized(callback) {
   let request = SpecialPowers._getQuotaManager().temporaryStorageInitialized();
   request.callback = callback;
@@ -163,6 +170,13 @@
   return request;
 }
 
+function initializePersistentStorage(callback) {
+  let request = SpecialPowers._getQuotaManager().initializePersistentStorage();
+  request.callback = callback;
+
+  return request;
+}
+
 function initTemporaryStorage(callback) {
   let request = SpecialPowers._getQuotaManager().initTemporaryStorage();
   request.callback = callback;
@@ -665,8 +679,13 @@
 
 async function verifyInitializationStatus(
   expectStorageIsInitialized,
+  expectPersistentStorageIsInitialized,
   expectTemporaryStorageIsInitialized
 ) {
+  if (!expectStorageIsInitialized && expectPersistentStorageIsInitialized) {
+    throw new Error("Invalid expectation");
+  }
+
   if (!expectStorageIsInitialized && expectTemporaryStorageIsInitialized) {
     throw new Error("Invalid expectation");
   }
@@ -676,12 +695,22 @@
 
   const storageIsInitialized = request.result;
 
+  request = persistentStorageInitialized();
+  await requestFinished(request);
+
+  const persistentStorageIsInitialized = request.result;
+
   request = temporaryStorageInitialized();
   await requestFinished(request);
 
   const temporaryStorageIsInitialized = request.result;
 
   ok(
+    !(!storageIsInitialized && persistentStorageIsInitialized),
+    "Initialization status is consistent"
+  );
+
+  ok(
     !(!storageIsInitialized && temporaryStorageIsInitialized),
     "Initialization status is consistent"
   );
@@ -692,6 +721,15 @@
     ok(!storageIsInitialized, "Storage is not initialized");
   }
 
+  if (expectPersistentStorageIsInitialized) {
+    ok(persistentStorageIsInitialized, "Persistent storage is initialized");
+  } else {
+    ok(
+      !persistentStorageIsInitialized,
+      "Persistent storage is not initialized"
+    );
+  }
+
   if (expectTemporaryStorageIsInitialized) {
     ok(temporaryStorageIsInitialized, "Temporary storage is initialized");
   } else {